### PR TITLE
RT-1.4 Adding Native SRL ACL compatibility

### PR DESCRIPTION
**Pull request by MarcCharlebois**
- Added `configACLNative()`, `configAdmitAllACLNative()`, `configACLInterfaceNative()` for ACL configurations on native SRL model
- Added a Deviation flag `deviation_use_native_acl_config` to enable using the above mentioned functions

This code is a Contribution to the OpenConfig Feature Profiles project ("Work") made under the Google Software Grant and Corporate Contributor License Agreement ("CLA") and governed by the Apache License 2.0. No other rights or licenses in or to any of Nokia’s intellectual property are granted for any other purpose. This code is provided on an "as is" basis without any warranties of any kind.

### CODE REVIEW

File: feature/bgp/gracefulrestart/ate_tests/bgp_graceful_restart_test/bgp_graceful_restart_test.go [hunk 1]
```diff
@@ -460,6 +463,217 @@ func configACLInterface(t *testing.T, iFace *oc.Acl_Interface, ifName string) *a
 	return aclConf
 }
 
+// Helper function to replicate configACL() configs in NOKIA native model
```
Comment (xw-g) [inline feature/bgp/gracefulrestart/ate_tests/bgp_graceful_restart_test/bgp_graceful_restart_test.go:466]: Remove the Nokia keyword.  "in vendor native/CLI model".

The func has a internal switch on vendors. The comment should not just mention Nokia.

File: feature/bgp/gracefulrestart/ate_tests/bgp_graceful_restart_test/bgp_graceful_restart_test.go [hunk 2]
```diff
@@ -460,6 +463,217 @@ func configACLInterface(t *testing.T, iFace *oc.Acl_Interface, ifName string) *a
...
+					},
+				},
+			},
+		}
+		gnmiClient := d.RawAPIs().GNMI().Default(t)
+		if _, err := gnmiClient.Set(context.Background(), gpbSetRequest); err != nil {
+			t.Fatalf("Unexpected error configuring SRL ACL: %v", err)
+		}
+	default:
+		t.Fatalf("Got vendor: %s, want: NOKIA", d.Vendor())
```
Comment (xw-g) [inline feature/bgp/gracefulrestart/ate_tests/bgp_graceful_restart_test/bgp_graceful_restart_test.go:587]: t.Fatalf("unsupported vendor, blablabla").

File: feature/bgp/gracefulrestart/ate_tests/bgp_graceful_restart_test/bgp_graceful_restart_test.go [hunk 3]
```diff
@@ -460,6 +463,217 @@ func configACLInterface(t *testing.T, iFace *oc.Acl_Interface, ifName string) *a
...
+	}
+}
+
+// Helper function to replicate AdminAllACL() configs in NOKIA native model
```
Comment (xw-g) [inline feature/bgp/gracefulrestart/ate_tests/bgp_graceful_restart_test/bgp_graceful_restart_test.go:591]: same comment as above.

File: feature/bgp/gracefulrestart/ate_tests/bgp_graceful_restart_test/bgp_graceful_restart_test.go [hunk 4]
```diff
@@ -460,6 +463,217 @@ func configACLInterface(t *testing.T, iFace *oc.Acl_Interface, ifName string) *a
...
+					},
+				},
+			},
+		}
+		gnmiClient := d.RawAPIs().GNMI().Default(t)
+		if _, err := gnmiClient.Set(context.Background(), gpbDelRequest); err != nil {
+			t.Fatalf("Unexpected error removing SRL ACL: %v", err)
+		}
+	default:
+		t.Fatalf("Got vendor: %s, want: NOKIA", d.Vendor())
```
Comment (xw-g) [inline feature/bgp/gracefulrestart/ate_tests/bgp_graceful_restart_test/bgp_graceful_restart_test.go:624]: Same comment as above.

File: feature/bgp/gracefulrestart/ate_tests/bgp_graceful_restart_test/bgp_graceful_restart_test.go [hunk 5]
```diff
@@ -460,6 +463,217 @@ func configACLInterface(t *testing.T, iFace *oc.Acl_Interface, ifName string) *a
...
+					},
+				},
+			},
+		}
+		gnmiClient := d.RawAPIs().GNMI().Default(t)
+		if _, err := gnmiClient.Set(context.Background(), gpbSetRequest); err != nil {
+			t.Fatalf("Unexpected error configuring interface ACL: %v", err)
+		}
+	default:
+		t.Fatalf("Got vendor: %s, want: NOKIA", d.Vendor())
```
Comment (xw-g) [inline feature/bgp/gracefulrestart/ate_tests/bgp_graceful_restart_test/bgp_graceful_restart_test.go:673]: Same comment as above.

File: feature/bgp/gracefulrestart/ate_tests/bgp_graceful_restart_test/bgp_graceful_restart_test.go [hunk 6]
```diff
@@ -508,14 +722,19 @@ func TestTrafficWithGracefulRestartSpeaker(t *testing.T) {
 	t.Run("VerifyTrafficPasswithGRTimerWithAclApplied", func(t *testing.T) {
 		t.Log("Configure Acl to block BGP on port 179")
 		const stopDuration = 45 * time.Second
-		gnmi.Replace(t, dut, gnmi.OC().Acl().AclSet(aclName, oc.Acl_ACL_TYPE_ACL_IPV4).Config(), configACL(d, aclName))
-		aclConf := configACLInterface(t, iFace, ifName)
+		if !*deviations.UseNativeACLConfiguration {
```
Comment (xw-g) [inline feature/bgp/gracefulrestart/ate_tests/bgp_graceful_restart_test/bgp_graceful_restart_test.go:725]: I think it's more natural to use the positive condition here. 

if true {
     configACLNative(t, dut, aclName)
     configACLInterfaceNative(t, dut, ifName)
} else {
     the_normal_funcs()
}

File: feature/bgp/gracefulrestart/ate_tests/bgp_graceful_restart_test/bgp_graceful_restart_test.go [hunk 7]
```diff
@@ -556,9 +775,14 @@ func TestTrafficWithGracefulRestartSpeaker(t *testing.T) {
 
 	t.Run("RemoveAclInterface", func(t *testing.T) {
 		t.Log("Removing Acl on the interface to restore BGP GR. Traffic should now pass!")
-		gnmi.Replace(t, dut, gnmi.OC().Acl().AclSet(aclName, oc.Acl_ACL_TYPE_ACL_IPV4).Config(), configAdmitAllACL(d, aclName))
-		aclPath := configACLInterface(t, iFace, ifName)
-		gnmi.Replace(t, dut, aclPath.Config(), iFace)
+		if !*deviations.UseNativeACLConfiguration {
```
Comment (xw-g) [inline feature/bgp/gracefulrestart/ate_tests/bgp_graceful_restart_test/bgp_graceful_restart_test.go:778]: same as above.

File: internal/deviations/deviations.go [hunk 8]
```diff
@@ -231,4 +231,6 @@ var (
 	MacAddressMissing = flag.Bool("deviation_mac_address_missing", false, "Device does not support /system/mac-address/state.")
 
 	gribiMACOverrideWithStaticARP = flag.Bool("deviation_gribi_mac_override_with_static_arp", false, "Set to true for device not supporting programming a gribi flow with a next-hop entry of mac-address only, default is false")
+
+	UseNativeACLConfiguration = flag.Bool("deviation_use_native_acl_config", false, "Configure ACLs using native model specifically for RT-1.4")
```
Comment (xw-g) [inline internal/deviations/deviations.go:235]: "UseVendorNativeACLConfiguration". 

"Configure ACLs using vendor native model specifically for RT-1.4".

**Comment by xw-g** [comment]
b/277130090 tracks the internal deviation.

File: internal/deviations/deviations.go (hunk 8 above)
Comment (MarcCharlebois) [reply internal/deviations/deviations.go]: sorry missed changing this when I fetched upstream, fixed now

File: feature/bgp/gracefulrestart/ate_tests/bgp_graceful_restart_test/bgp_graceful_restart_test.go [hunk 9]
```diff
@@ -460,6 +463,217 @@ func configACLInterface(t *testing.T, iFace *oc.Acl_Interface, ifName string) *a
...
+					},
+				},
+			},
+		}
+		gnmiClient := d.RawAPIs().GNMI().Default(t)
+		if _, err := gnmiClient.Set(context.Background(), gpbSetRequest); err != nil {
+			t.Fatalf("Unexpected error configuring SRL ACL: %v", err)
+		}
+	default:
+		t.Fatalf("Unsupported vendor, got vendor: %s, want: NOKIA", d.Vendor())
```
Comment (xw-g) [inline feature/bgp/gracefulrestart/ate_tests/bgp_graceful_restart_test/bgp_graceful_restart_test.go:587]: nit: remove "want: NOKIA".

File: feature/bgp/gracefulrestart/ate_tests/bgp_graceful_restart_test/bgp_graceful_restart_test.go [hunk 10]
```diff
@@ -460,6 +463,217 @@ func configACLInterface(t *testing.T, iFace *oc.Acl_Interface, ifName string) *a
...
+	}
+}
+
+// Helper function to replicate AdminAllACL() configs in native model
+// Remove ACL entries 10, 20
```
Comment (xw-g) [inline feature/bgp/gracefulrestart/ate_tests/bgp_graceful_restart_test/bgp_graceful_restart_test.go:592]: should it be this? "Helper function to replicate AdminAllACL() configs in native model, but remove ACL entries 10, 20"

File: feature/bgp/gracefulrestart/ate_tests/bgp_graceful_restart_test/bgp_graceful_restart_test.go [hunk 11]
```diff
@@ -460,6 +463,217 @@ func configACLInterface(t *testing.T, iFace *oc.Acl_Interface, ifName string) *a
...
+					},
+				},
+			},
+		}
+		gnmiClient := d.RawAPIs().GNMI().Default(t)
+		if _, err := gnmiClient.Set(context.Background(), gpbDelRequest); err != nil {
+			t.Fatalf("Unexpected error removing SRL ACL: %v", err)
+		}
+	default:
+		t.Fatalf("Unsupported vendor, got vendor: %s, want: NOKIA", d.Vendor())
```
Comment (xw-g) [inline feature/bgp/gracefulrestart/ate_tests/bgp_graceful_restart_test/bgp_graceful_restart_test.go:624]: same, remove "want: NOKIA".

File: feature/bgp/gracefulrestart/ate_tests/bgp_graceful_restart_test/bgp_graceful_restart_test.go [hunk 12]
```diff
@@ -460,6 +463,217 @@ func configACLInterface(t *testing.T, iFace *oc.Acl_Interface, ifName string) *a
...
+	}
+}
+
+// Helper function to replicate configACLInterface in native model
```
Comment (xw-g) [inline feature/bgp/gracefulrestart/ate_tests/bgp_graceful_restart_test/bgp_graceful_restart_test.go:628]: Seems like comment synax is missing comma and period? Note that Golang comment has no length requirement/restriction.

File: feature/bgp/gracefulrestart/ate_tests/bgp_graceful_restart_test/bgp_graceful_restart_test.go [hunk 13]
```diff
@@ -460,6 +463,217 @@ func configACLInterface(t *testing.T, iFace *oc.Acl_Interface, ifName string) *a
...
+					},
+				},
+			},
+		}
+		gnmiClient := d.RawAPIs().GNMI().Default(t)
+		if _, err := gnmiClient.Set(context.Background(), gpbSetRequest); err != nil {
+			t.Fatalf("Unexpected error configuring interface ACL: %v", err)
+		}
+	default:
+		t.Fatalf("Unsupported vendor, got vendor: %s, want: NOKIA", d.Vendor())
```
Comment (xw-g) [inline feature/bgp/gracefulrestart/ate_tests/bgp_graceful_restart_test/bgp_graceful_restart_test.go:673]: Remove "want: NOKIA".

File: feature/bgp/gracefulrestart/ate_tests/bgp_graceful_restart_test/bgp_graceful_restart_test.go (hunk 10 above)
Comment (MarcCharlebois) [reply feature/bgp/gracefulrestart/ate_tests/bgp_graceful_restart_test/bgp_graceful_restart_test.go:592]: Yes it should, will fix typo!

File: feature/bgp/gracefulrestart/ate_tests/bgp_graceful_restart_test/bgp_graceful_restart_test.go [hunk 14]
```diff
@@ -508,14 +722,19 @@ func TestTrafficWithGracefulRestartSpeaker(t *testing.T) {
...
+			gnmi.Replace(t, dut, gnmi.OC().Acl().AclSet(aclName, oc.Acl_ACL_TYPE_ACL_IPV4).Config(), configACL(d, aclName))
+			aclConf := configACLInterface(t, iFace, ifName)
+			gnmi.Replace(t, dut, aclConf.Config(), iFace)
+		}
 		t.Log("Starting traffic")
 		ate.Traffic().Start(t, allFlows...)
 		startTime := time.Now()
 		t.Log("Trigger Graceful Restart on ATE")
 		ate.Actions().NewBGPGracefulRestart().WithRestartTime(grRestartTime).WithPeers(bgpPeer).Send(t)
-		gnmi.Replace(t, dut, aclConf.Config(), iFace)
```
Comment (xw-g) [inline feature/bgp/gracefulrestart/ate_tests/bgp_graceful_restart_test/bgp_graceful_restart_test.go:518]: With this changes now we started blocking BGP packet even before `ate.Actions().NewBGPGracefulRestart().WithRestartTime(grRestartTime).WithPeers(bgpPeer).Send(t)`?

**Comment by xw-g** [comment]
@cprabha Seems like there is some issues with our internal presubmit testbed. Can you help check if your team is has any concern with this? If not, can we go ahead and merge? thanks.

**Comment by cprabha** [comment]
> d merge? thanks.

It looks good. Thanks.

**Review by cprabha — APPROVED** [review]

**Comment by xw-g** [comment]
b/277367138 records the presubmit effort.

**Review by xw-g — APPROVED** [review]

**Pull request merged**